src/som/interpreter/SArguments.java

```diff
@@ -58,4 +91,139 @@ public static SImmutableArray getArgumentsWithoutReceiver(final Object[] argumen
+      final VirtualFrame frame) {
+    if (VmSettings.ACTOR_ASYNC_STACK_TRACE_STRUCTURE) {
+      Object[] arguments = new Object[] {receiver, firstArg, null};
+      entryLoad.loadShadowStackEntry(arguments, expression, frame, false);
+      return arguments;
+    } else {
+      return new Object[] {receiver, firstArg};
+    }
+  }
+
+  public static Object[] getPlain2ArgumentsWithReceiver(final Object receiver,
+      final Object firstArg,
+      final Object secondArg,
+      final ExpressionNode expression,
+      final ShadowStackEntryLoad entryLoad,
+      final VirtualFrame frame) {
+    if (VmSettings.ACTOR_ASYNC_STACK_TRACE_STRUCTURE) {
+      Object[] arguments = new Object[] {receiver, firstArg, secondArg, null};
+      entryLoad.loadShadowStackEntry(arguments, expression, frame, false);
+      return arguments;
+    } else {
+      return new Object[] {receiver, firstArg, secondArg};
+    }
+  }
+
+  public static Object[] getPlainArguments(final Object[] args) {
+    if (VmSettings.ACTOR_ASYNC_STACK_TRACE_STRUCTURE) {
+      Object[] newArgs = new Object[args.length + 1];
+      for (int i = 0; i < args.length; i++) {
+        newArgs[i] = args[i];
+      }
+      return newArgs;
+    } else {
+      return args;
+    }
+  }
+
+  public static Object[] getPromiseCallbackArgumentArray(final SBlock callback) {
+    if (VmSettings.ACTOR_ASYNC_STACK_TRACE_STRUCTURE) {
+      return new Object[] {callback, null, null};
+    } else {
+      return new Object[] {callback, null};
+    }
+  }
+
+  public static void setShadowStackEntryWithCache(final Object[] arguments,
+      final Node expression,
+      final ShadowStackEntryLoad entryLoad,
+      final VirtualFrame frame,
+      final boolean async) {
+    if (VmSettings.ACTOR_ASYNC_STACK_TRACE_STRUCTURE) {
+      entryLoad.loadShadowStackEntry(arguments, expression, frame, async);
+    }
+  }
+
+  public static ShadowStackEntry instantiateTopShadowStackEntry(final Node expr) {
+    return ShadowStackEntry.createTop(expr);
+  }
+
+  public static ShadowStackEntry instantiateShadowStackEntry(final ShadowStackEntry previous,
+      final Node expr, final boolean async) {
+    if (async) {
+      return ShadowStackEntry.createAtAsyncSend(previous, expr);
+    } else {
+      return ShadowStackEntry.create(previous, expr);
+    }
+  }
+
+  public static ShadowStackEntry getShadowStackEntry(final VirtualFrame frame) {
+    Object[] args = frame.getArguments();
+    return getShadowStackEntry(args);
+  }
+
+  public static ShadowStackEntry getShadowStackEntry(final Object[] args) {
+    Object maybeShadowStack = args[args.length - 1];
+    if (maybeShadowStack instanceof ShadowStackEntry) {
+      return (ShadowStackEntry) maybeShadowStack;
+    }
+    return null;
+  }
+
+  public static void setShadowStackEntry(final Object[] args,
+      final ShadowStackEntry entry) {
+    // assert args[args.length - 1] == null : "Assume shadow stack entry is not already set.";
```

Comment:
Hm, why is this commented out?
Seems like a sensible assertion, or does it not hold? If so, in which case does it fail, should the assertion be removed or documented?

Reply:
It should be uncommented. Done.

Reply:
@ctrlpz please do not commit such a change. These assertions are there for a reason. It just cost me a lot of time that the assertion wasn't there.